File: FashionStore/ViewController.m
//
//  ViewController.m
//  FashionStore
//
//  Copyright © 2016 Klarna. All rights reserved.
//

#import "ViewController.h"
#import <KlarnaCheckoutSDK/KlarnaCheckout.h>

@interface ViewController ()

@property (nonatomic, weak) IBOutlet UIWebView *webView;
@property (nonatomic, strong) KCOKlarnaCheckout *checkout;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    [self addObservers];
    self.title = @"iOS SDK Sample App";
    self.webView.keyboardDisplayRequiresUserAction = NO;
    self.checkout = [[KCOKlarnaCheckout alloc] initWithViewController:self returnURL:[NSURL URLWithString:@"fashionstore://"]];
    [self.checkout setWebView:self.webView];
    [self.checkout notifyViewDidLoad];

    NSURL *url = [NSURL URLWithString:@"https://www.klarna.com/demo/"];
    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    [self.webView loadRequest:request];
}

#pragma mark - Internals

- (void)addObservers
{
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleNotification:) name:KCOSignalNotification object:nil];
}

- (void)removeObservers
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)handleNotification:(NSNotification *)notification
{
    NSString *name = notification.userInfo[KCOSignalNameKey];
    NSDictionary *data = notification.userInfo[KCOSignalDataKey];

    if ([name isEqualToString:@"complete"]) {
        [self handleCompletionUri:[data objectForKey:@"uri"]];
    }
}

- (void)handleCompletionUri:(NSString *)uri{
    if (uri && [uri isKindOfClass:[NSString class]] && uri.length > 0) {
        NSURL *url = [NSURL URLWithString:uri];
        [self.webView loadRequest:[NSURLRequest requestWithURL:url]];
    }
}

@end

File: KlarnaCheckoutExampleSwift/KlarnaCheckoutExampleSwift/ViewController.swift
import UIKit

class ViewController: UIViewController {
    
    @IBOutlet weak var checkoutWebview: UIWebView!
    @IBOutlet weak var confirmationWebView: UIWebView!
    var checkout: KCOKlarnaCheckout?
    var snippetLoader: KCOSnippetLoader?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        checkout = KCOKlarnaCheckout(viewController: self)
        checkout?.setWebView(checkoutWebview)
        snippetLoader = KCOSnippetLoader(webview: checkoutWebview)
        snippetLoader?.delegate = self
        
        NotificationCenter.default.addObserver(self, selector: #selector(handleNotification), name: NSNotification.Name.KCOSignal, object: nil)
        checkout?.notifyViewDidLoad()
        
        loadCheckoutUrl()
    }
    
    fileprivate func loadCheckoutUrl() {
        guard let url = URL(string: "https://www.klarnacheckout.com") else { return }
        let request = URLRequest(url: url)
        checkoutWebview.loadRequest(request)
    }
    
    func handleNotification(_ notification: Notification) {
        let name = notification.userInfo?[KCOSignalNameKey] as? String ?? ""
        let data = notification.userInfo?[KCOSignalDataKey] as? [AnyHashable: Any] ?? [:]
        
        if name == "complete" {
            if let dataDict = data[0] as? [String: AnyObject], let uriString = dataDict["uri"] as? String {
                handleCompletionUri(uriString)
            }
        }
    }
    
    fileprivate func handleCompletionUri(_ uriString: String) {
        if uriString != "" {
            if let url = URL(string: uriString) {
                let request = URLRequest(url: url)
                checkout?.setWebView(confirmationWebView)
                confirmationWebView.loadRequest(request)
            }
        }
    }
}

extension ViewController: KCOSnippetLoaderDelegate {
    func snippetLoader(snippetLoader: KCOSnippetLoader, loadedCheckout version: KCOCheckoutInfo) {
        //checkout?.setSnippet(version.snippet)
    }
}
